File: HypnoScript.Core/Symbols/SymbolTable.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace HypnoScript.Core.Symbols
{
	// Runtime-Level: Erweiterte SymbolTable mit Debugging und Scope-Analyse
	public class SymbolTable
	{
		private readonly SymbolTable? _enclosing;
		private readonly Dictionary<string, Symbol> _symbols = new();
		private readonly List<SymbolTable> _childScopes = new();
		public string ScopeName { get; set; } = "Global";
		public int ScopeLevel { get; }

		public SymbolTable(SymbolTable? enclosing = null, string scopeName = "Global")
		{
			_enclosing = enclosing;
			ScopeName = scopeName;
			ScopeLevel = enclosing?.ScopeLevel + 1 ?? 0;
			enclosing?._childScopes.Add(this);
		}

		public bool Define(Symbol sym)
		{
			if (_symbols.ContainsKey(sym.Name))
			{
				Console.Error.WriteLine($"[SymbolTable] Symbol '{sym.Name}' is already defined in scope '{ScopeName}'.");
				return false;
			}
			_symbols[sym.Name] = sym;
			return true;
		}

		public Symbol? Resolve(string name)
		{
			if (_symbols.TryGetValue(name, out var sym))
				return sym;
			return _enclosing?.Resolve(name);
		}

		public Symbol? ResolveLocal(string name)
		{
			_symbols.TryGetValue(name, out var sym);
			return sym;
		}

		public bool Assign(string name, object? value)
		{
			var symbol = Resolve(name);
			if (symbol == null)
			{
				Console.Error.WriteLine($"[SymbolTable] Cannot assign to undefined symbol '{name}'.");
				return false;
			}
			if (symbol.IsConstant)
			{
				Console.Error.WriteLine($"[SymbolTable] Cannot assign to constant symbol '{name}'.");
				return false;
			}
			symbol.Value = value;
			return true;
		}

		// Runtime-Level: Methode, um den aktuellen Scope-Stack als String auszugeben
		public string DebugScope()
		{
			var result = $"Scope '{ScopeName}' (Level {ScopeLevel}):\n";
			foreach (var kvp in _symbols.OrderBy(x => x.Key))
			{
				var symbol = kvp.Value;
				var valueInfo = symbol.Value != null ? $" = {symbol.Value}" : "";
				var constInfo = symbol.IsConstant ? " (const)" : "";
				var exportInfo = symbol.IsExported ? " (exported)" : "";
				result += $"  {symbol.Kind} {kvp.Key}: {symbol.TypeName}{valueInfo}{constInfo}{exportInfo}\n";
			}
			if (_enclosing != null)
			{
				result += "\nEnclosing Scope:\n" + _enclosing.DebugScope();
			}
			return result;
		}

		// Neue Runtime-Features
		public IEnumerable<Symbol> GetAllSymbols()
		{
			return _symbols.Values.OrderBy(s => s.Name);
		}

		public IEnumerable<Symbol> GetSymbolsByKind(SymbolKind kind)
		{
			return _symbols.Values.Where(s => s.Kind == kind).OrderBy(s => s.Name);
		}

		public IEnumerable<Symbol> GetExportedSymbols()
		{
			return _symbols.Values.Where(s => s.IsExported).OrderBy(s => s.Name);
		}

		public IEnumerable<Symbol> GetConstants()
		{
			return _symbols.Values.Where(s => s.IsConstant).OrderBy(s => s.Name);
		}

		public int SymbolCount => _symbols.Count;

		public bool HasSymbol(string name)
		{
			return _symbols.ContainsKey(name);
		}

		public bool RemoveSymbol(string name)
		{
			return _symbols.Remove(name);
		}

		public void Clear()
		{
			_symbols.Clear();
		}

		// Scope-Hierarchie-Management
		public SymbolTable? GetEnclosingScope() => _enclosing;
		public IEnumerable<SymbolTable> GetChildScopes() => _childScopes;

		public SymbolTable GetRootScope()
		{
			var current = this;
			while (current._enclosing != null)
			{
				current = current._enclosing;
			}
			return current;
		}

		public int GetScopeDepth()
		{
			var depth = 0;
			var current = this;
			while (current._enclosing != null)
			{
				depth++;
				current = current._enclosing;
			}
			return depth;
		}

		// Symbol-Statistiken
		public Dictionary<SymbolKind, int> GetSymbolStatistics()
		{
			return _symbols.Values
				.GroupBy(s => s.Kind)
				.ToDictionary(g => g.Key, g => g.Count());
		}

		public string GetScopeSummary()
		{
			var stats = GetSymbolStatistics();
			var summary = $"Scope '{ScopeName}' (Level {ScopeLevel}): {SymbolCount} symbols\n";
			foreach (var stat in stats.OrderBy(s => s.Key))
			{
				summary += $"  {stat.Key}: {stat.Value}\n";
			}
			return summary;
		}

		// Symbol-Suche mit Filter
		public IEnumerable<Symbol> SearchSymbols(string pattern, SymbolKind? kind = null)
		{
			var query = _symbols.Values.AsEnumerable();

			if (kind.HasValue)
				query = query.Where(s => s.Kind == kind.Value);

			return query.Where(s => s.Name.Contains(pattern, StringComparison.OrdinalIgnoreCase))
					   .OrderBy(s => s.Name);
		}

		// Symbol-Validierung
		public List<string> ValidateSymbols()
		{
			var errors = new List<string>();

			foreach (var symbol in _symbols.Values)
			{
				if (string.IsNullOrWhiteSpace(symbol.Name))
					errors.Add($"Symbol has empty name in scope '{ScopeName}'");

				if (symbol.Kind == SymbolKind.Function && string.IsNullOrEmpty(symbol.TypeName))
					errors.Add($"Function '{symbol.Name}' has no return type");

				if (symbol.IsConstant && symbol.Value == null)
					errors.Add($"Constant '{symbol.Name}' has no initial value");
			}

			return errors;
		}

		// Scope-Merging (für Module/Imports)
		public void MergeFrom(SymbolTable other, bool overwrite = false)
		{
			foreach (var kvp in other._symbols)
			{
				if (overwrite || !_symbols.ContainsKey(kvp.Key))
				{
					_symbols[kvp.Key] = kvp.Value;
				}
			}
		}

		// Scope-Export (für Module)
		public SymbolTable ExportScope()
		{
			var exported = new SymbolTable(null, $"{ScopeName}_Exported");
			foreach (var symbol in _symbols.Values.Where(s => s.IsExported))
			{
				exported.Define(symbol);
			}
			return exported;
		}
	}
}

File: hypnoscript-core/src/symbol_table.rs
use crate::symbols::{Symbol, SymbolKind};
use std::collections::HashMap;

/// Symbol table for managing scopes and variable bindings
#[derive(Debug, Clone)]
pub struct SymbolTable {
    enclosing: Option<Box<SymbolTable>>,
    symbols: HashMap<String, Symbol>,
    child_scopes: Vec<SymbolTable>,
    pub scope_name: String,
    pub scope_level: usize,
}

impl SymbolTable {
    /// Create a new symbol table
    pub fn new(enclosing: Option<Box<SymbolTable>>, scope_name: String) -> Self {
        let scope_level = enclosing.as_ref().map(|e| e.scope_level + 1).unwrap_or(0);
        Self {
            enclosing,
            symbols: HashMap::new(),
            child_scopes: Vec::new(),
            scope_name,
            scope_level,
        }
    }

    /// Create a global scope
    pub fn global() -> Self {
        Self::new(None, "Global".to_string())
    }

    /// Define a new symbol in the current scope
    pub fn define(&mut self, sym: Symbol) -> bool {
        if self.symbols.contains_key(&sym.name) {
            eprintln!(
                "[SymbolTable] Symbol '{}' is already defined in scope '{}'.",
                sym.name, self.scope_name
            );
            return false;
        }
        self.symbols.insert(sym.name.clone(), sym);
        true
    }

    /// Resolve a symbol, looking in enclosing scopes if necessary
    pub fn resolve(&self, name: &str) -> Option<&Symbol> {
        self.symbols
            .get(name)
            .or_else(|| self.enclosing.as_ref().and_then(|e| e.resolve(name)))
    }

    /// Resolve a symbol only in the current scope
    pub fn resolve_local(&self, name: &str) -> Option<&Symbol> {
        self.symbols.get(name)
    }

    /// Check if a symbol exists (locally or in enclosing scopes)
    pub fn has_symbol(&self, name: &str) -> bool {
        self.resolve(name).is_some()
    }

    /// Remove a symbol from the current scope
    pub fn remove_symbol(&mut self, name: &str) -> bool {
        self.symbols.remove(name).is_some()
    }

    /// Clear all symbols from the current scope
    pub fn clear(&mut self) {
        self.symbols.clear();
    }

    /// Get all symbols in the current scope
    pub fn get_all_symbols(&self) -> Vec<&Symbol> {
        let mut symbols: Vec<_> = self.symbols.values().collect();
        symbols.sort_by(|a, b| a.name.cmp(&b.name));
        symbols
    }

    /// Get symbols by kind
    pub fn get_symbols_by_kind(&self, kind: SymbolKind) -> Vec<&Symbol> {
        let mut symbols: Vec<_> = self.symbols.values().filter(|s| s.kind == kind).collect();
        symbols.sort_by(|a, b| a.name.cmp(&b.name));
        symbols
    }

    /// Get exported symbols
    pub fn get_exported_symbols(&self) -> Vec<&Symbol> {
        let mut symbols: Vec<_> = self.symbols.values().filter(|s| s.is_exported).collect();
        symbols.sort_by(|a, b| a.name.cmp(&b.name));
        symbols
    }

    /// Get constants
    pub fn get_constants(&self) -> Vec<&Symbol> {
        let mut symbols: Vec<_> = self.symbols.values().filter(|s| s.is_constant).collect();
        symbols.sort_by(|a, b| a.name.cmp(&b.name));
        symbols
    }

    /// Get symbol count
    pub fn symbol_count(&self) -> usize {
        self.symbols.len()
    }

    /// Get the enclosing scope
    pub fn get_enclosing_scope(&self) -> Option<&SymbolTable> {
        self.enclosing.as_ref().map(|e| e.as_ref())
    }

    /// Get child scopes
    pub fn get_child_scopes(&self) -> &[SymbolTable] {
        &self.child_scopes
    }

    /// Get the root scope
    pub fn get_root_scope(&self) -> &SymbolTable {
        let mut current = self;
        while let Some(ref enclosing) = current.enclosing {
            current = enclosing;
        }
        current
    }

    /// Get scope depth
    pub fn get_scope_depth(&self) -> usize {
        let mut depth = 0;
        let mut current = self;
        while let Some(ref enclosing) = current.enclosing {
            depth += 1;
            current = enclosing;
        }
        depth
    }

    /// Get symbol statistics
    pub fn get_symbol_statistics(&self) -> HashMap<SymbolKind, usize> {
        let mut stats = HashMap::new();
        for symbol in self.symbols.values() {
            *stats.entry(symbol.kind).or_insert(0) += 1;
        }
        stats
    }

    /// Get a scope summary
    pub fn get_scope_summary(&self) -> String {
        let stats = self.get_symbol_statistics();
        let mut summary = format!(
            "Scope '{}' (Level {}): {} symbols\n",
            self.scope_name,
            self.scope_level,
            self.symbol_count()
        );

        let mut kinds: Vec<_> = stats.keys().collect();
        kinds.sort();
        for kind in kinds {
            if let Some(count) = stats.get(kind) {
                summary.push_str(&format!("  {:?}: {}\n", kind, count));
            }
        }
        summary
    }

    /// Search symbols by pattern
    pub fn search_symbols(&self, pattern: &str, kind: Option<SymbolKind>) -> Vec<&Symbol> {
        let pattern_lower = pattern.to_lowercase();
        let mut symbols: Vec<_> = self
            .symbols
            .values()
            .filter(|s| {
                let name_match = s.name.to_lowercase().contains(&pattern_lower);
                let kind_match = kind.is_none_or(|k| s.kind == k);
                name_match && kind_match
            })
            .collect();
        symbols.sort_by(|a, b| a.name.cmp(&b.name));
        symbols
    }

    /// Validate symbols
    pub fn validate_symbols(&self) -> Vec<String> {
        let mut errors = Vec::new();

        for symbol in self.symbols.values() {
            if symbol.name.trim().is_empty() {
                errors.push(format!(
                    "Symbol has empty name in scope '{}'",
                    self.scope_name
                ));
            }

            if symbol.kind == SymbolKind::Function && symbol.type_name.is_none() {
                errors.push(format!("Function '{}' has no return type", symbol.name));
            }
        }

        errors
    }

    /// Merge symbols from another symbol table
    pub fn merge_from(&mut self, other: &SymbolTable, overwrite: bool) {
        for (name, symbol) in &other.symbols {
            if overwrite || !self.symbols.contains_key(name) {
                self.symbols.insert(name.clone(), symbol.clone());
            }
        }
    }

    /// Export only exported symbols to a new scope
    pub fn export_scope(&self) -> SymbolTable {
        let mut exported = SymbolTable::new(None, format!("{}_Exported", self.scope_name));
        for symbol in self.symbols.values() {
            if symbol.is_exported {
                exported.define(symbol.clone());
            }
        }
        exported
    }

    /// Debug scope information
    pub fn debug_scope(&self) -> String {
        let mut result = format!(
            "Scope '{}' (Level {}):\n",
            self.scope_name, self.scope_level
        );

        let mut symbols: Vec<_> = self.symbols.iter().collect();
        symbols.sort_by(|a, b| a.0.cmp(b.0));

        for (name, symbol) in symbols {
            let const_info = if symbol.is_constant { " (const)" } else { "" };
            let export_info = if symbol.is_exported {
                " (exported)"
            } else {
                ""
            };
            result.push_str(&format!(
                "  {:?} {}: {:?}{}{}\n",
                symbol.kind, name, symbol.type_name, const_info, export_info
            ));
        }

        if let Some(ref enclosing) = self.enclosing {
            result.push_str("\nEnclosing Scope:\n");
            result.push_str(&enclosing.debug_scope());
        }

        result
    }
}
